gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/filter/DedupGlobalStep.java

```diff
@@ -89,6 +92,17 @@ public ElementRequirement getMaxRequirement() {
 
     @Override
     protected Traverser.Admin<S> processNextStart() {
+        if (null != this.barrier) {
+            this.barrierIterator = this.barrier.entrySet().iterator();
+            this.barrier = null;
+        }
+        while (this.barrierIterator != null && this.barrierIterator.hasNext()) {
+            if (null == this.barrierIterator)
```

Comment:
Your code fails. Did you run it? ... The problem is that `this.barrier` is null'd but the `barrierIterator` still exists and you still need to fetch results from it.